proxy/src/main/java/com/wavefront/agent/TenantInfo.java

```diff
@@ -0,0 +1,196 @@
+                                                    long currentTime, long currentDuration) {
+                            return currentDuration;
+                        }
+                    }).build(key -> getAccessTokenByAPIToken());
+        } else {
+            // This step assumes that the authentication mechanism is WAVEFRONT_API_TOKEN.
+            tokenCache = null;
+        }
+    }
+
+    TenantInfo(@Nonnull final String serverToServiceClientId,
+               @Nonnull final String serverToServiceClientSecret,
+               @Nonnull final String orgId,
+               @Nonnull final String tenantServer) {
+        this.tenantServer = tenantServer;
+        this.tokenCache = Caffeine.newBuilder().
+                expireAfter(new Expiry<String, TokenExchangeResponseDTO>() {
+                    @Override
+                    public long expireAfterCreate(@Nonnull String authToken,
+                                                  @Nonnull TokenExchangeResponseDTO tokenExchangeResponseDTO,
+                                                  long currentTime) {
+                        return TimeUnit.SECONDS.toNanos(tokenExchangeResponseDTO.getExpiresIn());
+                    }
+
+                    @Override
+                    public long expireAfterUpdate(@Nonnull String authToken, @Nonnull TokenExchangeResponseDTO tokenExchangeResponseDTO,
+                                                  long currentTime, long currentDuration) {
+                        return currentDuration;
+                    }
+
+                    @Override
+                    public long expireAfterRead(@Nonnull String authToken, @Nonnull TokenExchangeResponseDTO tokenExchangeResponseDTO, long currentTime,
+                                                long currentDuration) {
+                        return currentDuration;
+                    }
+                }).build(key -> getAccessTokenByClientCredentials(serverToServiceClientId, serverToServiceClientSecret, orgId));
+    }
+
+    @Nonnull
+    public String getTenantServer() {
+        return tenantServer;
+    }
+
+    @Nonnull
+    public String getToken() {
+        if (tokenCache == null) {
+            // Returning Wavefront api token.
+            return tenantToken;
+        }
+
+        return tokenCache.get(CSP_AUTH_TOKEN).getAccessToken();
+    }
+
+    /**
+     * When the CSP access token has expired, obtain it using the CSP api token.
+     *
+     * @return the CSP token exchange dto.
+     * @throws RuntimeException when CSP is down or wrong CSP api token.
+     */
+    @Nonnull
+    private TokenExchangeResponseDTO getAccessTokenByAPIToken() {
+        Form body = new Form();
+        body.param("grant_type", "api_token");
+        body.param("api_token", tenantToken);
+
+        Response response =
+                new JerseyClientBuilder()
```

Comment:
use RestEasy, like other parts of the Proxy

Reply:
Done